controllers/awsmachinetemplate_controller.go

```diff
@@ -0,0 +1,376 @@
+	if len(capacity) > 0 {
+		awsMachineTemplate.Status.Capacity = capacity
+	}
+	if nodeInfo != nil && (nodeInfo.Architecture != "" || nodeInfo.OperatingSystem != "") {
+		awsMachineTemplate.Status.NodeInfo = nodeInfo
+	}
+	if err := r.Status().Patch(ctx, awsMachineTemplate, client.MergeFrom(original)); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to update AWSMachineTemplate status")
+	}
+
+	log.Info("Successfully populated capacity and nodeInfo", "instanceType", instanceType, "region", region, "capacity", capacity, "nodeInfo", nodeInfo)
+	return ctrl.Result{}, nil
+}
+
+// getRegion finds the region by checking the template's owner cluster reference.
+func (r *AWSMachineTemplateReconciler) getRegion(ctx context.Context, template *infrav1.AWSMachineTemplate) (string, error) {
+	// Get the owner cluster
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, template.ObjectMeta)
+	if err != nil {
+		return "", err
+	}
+	if cluster == nil {
+		return "", errors.New("no owner cluster found")
+	}
+
+	// Get region from AWSCluster (standard EC2-based cluster)
+	if cluster.Spec.InfrastructureRef != nil && cluster.Spec.InfrastructureRef.Kind == "AWSCluster" {
+		awsCluster := &infrav1.AWSCluster{}
+		if err := r.Get(ctx, client.ObjectKey{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Spec.InfrastructureRef.Name,
+		}, awsCluster); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return "", errors.Wrapf(err, "failed to get AWSCluster %s/%s", cluster.Namespace, cluster.Spec.InfrastructureRef.Name)
+			}
+		} else if awsCluster.Spec.Region != "" {
+			return awsCluster.Spec.Region, nil
+		}
+	}
+
+	return "", nil
+}
+
+// getInstanceTypeCapacity queries AWS EC2 API for instance type capacity information.
+// Returns the resource list (CPU, Memory).
+func (r *AWSMachineTemplateReconciler) getInstanceTypeCapacity(ctx context.Context, ec2Client *ec2.Client, instanceType string) (corev1.ResourceList, error) {
+	// Query instance type information
+	input := &ec2.DescribeInstanceTypesInput{
+		InstanceTypes: []ec2types.InstanceType{ec2types.InstanceType(instanceType)},
+	}
+
+	result, err := ec2Client.DescribeInstanceTypes(ctx, input)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to describe instance type %q", instanceType)
+	}
+
+	if len(result.InstanceTypes) == 0 {
+		return nil, errors.Errorf("no information found for instance type %q", instanceType)
+	}
+
+	// Extract capacity information
+	info := result.InstanceTypes[0]
+	resourceList := corev1.ResourceList{}
+
+	// CPU
+	if info.VCpuInfo != nil && info.VCpuInfo.DefaultVCpus != nil {
+		resourceList[corev1.ResourceCPU] = *resource.NewQuantity(int64(*info.VCpuInfo.DefaultVCpus), resource.DecimalSI)
+	}
+
+	// Memory
+	if info.MemoryInfo != nil && info.MemoryInfo.SizeInMiB != nil {
+		memoryBytes := *info.MemoryInfo.SizeInMiB * 1024 * 1024
+		resourceList[corev1.ResourceMemory] = *resource.NewQuantity(memoryBytes, resource.BinarySI)
+	}
```

Comment:
I'm not sure whether my handling of this part is correct, and I’m also uncertain about which unit I should be using.